_posts/2021-03-01-introducing-sql-delta-import.md

```diff
@@ -0,0 +1,144 @@
+number of output files and potentially introducing a small files problem.
+
+`sql delta import` uses `--chunks` parameter to control number of... well... chunks to split the source table 
+into and standard spark parameters like `--num-executors` and `--executor-cores` to control data import
+concurrency thus allowing you to tune those parameters independently
+
+```shell script
+spark-submit --num-executors 15 --executor-cores 4 /
+--conf spark.databricks.delta.optimizeWrite.enabled=true /
+--conf spark.databricks.delta.autoCompact.enabled=true /
+--class "com.scribd.importer.spark.ImportRunner" sql-delta-import_2.12-0.1.0-SNAPSHOT.jar /
+--jdbc-url jdbc:mysql://hostName:port/database /
+--source source.table
+--destination destination.table
+--split-by id
+--chunks 500
+```
+
+in the example above source table will be split into 500 chunks resulting in quick transactions and released connections
+but no more than 60 concurrent connections will be used for import since max degree of parallelism is 60 (15 executors x 4 cores). 
+`delta.optimizeWrite` and `delta.autoCompact` configuration will yield optimal file size output for the destination table
+
+#### 3.1 `--num-mappers` and data skew just don't play nicely together
+ 
+When `sqoop` imports data, source table will be split into ranges based on `--split-by` column and each mapper 
+would import it's corresponding range. This works good when `--split-by` column has a near uniform distribution 
+of data, but that's not always the case with source tables... As tables age we tend to add additional columns to them to
+take on new business requirements so over time data in latest rows has a higher fill rate than earlier rows. 
+
+![row density increase over time](/post-images/2021-03-sql-delta-import/row_density_increase.png)
+
+Our source tables here at Scribd definitely have these characteristics. We also have some tables that have entire 
+ranges of data missing due to data cleanup. At some point large chunks of data were just deleted from these tables.
+
+![missing rows](/post-images/2021-03-sql-delta-import/missing_rows.png)
+
+This type of data skew will result in processing time skew and output file size skew when you can only control number of 
+mappers. Yes we can introduce additional computed synthetic column in the source table as our `split-by` column but now 
+there is an additional column that does not add business value, app developers need to be aware of it, computing and 
+storing it takes up database resources and if we plan to use it for imports it's better be indexed, thus even more 
+compute and storage resources.
+
+With `sql-delta-import` we can "solve" this problem by making number of chunks much larger than max degree of parallelism.
+This way large chunks with high data density are broken up into smaller pieces that a single executor can handle. 
+Executors that get chunks with little or no data can just quickly process them and move on to do some real work. 
```

Comment:
We have mentioned how sqoop basically break data into chunks by split-by column for each mapper , but we havent mentioned that for sql-delta-import. Does it just break the entire table into chunks without even considering the split-by column?